provider.go

```diff
@@ -5,110 +5,178 @@ package azuresecrets
+}
+
+var _ AzureProvider = (*provider)(nil)
 
 // provider is a concrete implementation of AzureProvider. In most cases it is a simple passthrough
 // to the appropriate client object. But if the response requires processing that is more practical
 // at this layer, the response signature may different from the Azure signature.
 type provider struct {
 	settings *clientSettings
+	// HTTP Client used to make role assignment and definition requests
+	// using new arm libraries
+	httpClient *http.Client
 
 	appClient    api.ApplicationsClient
 	spClient     api.ServicePrincipalClient
 	groupsClient api.GroupsClient
-	raClient     *authorization.RoleAssignmentsClient
-	rdClient     *authorization.RoleDefinitionsClient
+	raClient     *armauthorization.RoleAssignmentsClient
+	rdClient     *armauthorization.RoleDefinitionsClient
 }
 
 // newAzureProvider creates an azureProvider, backed by Azure client objects for underlying services.
-func newAzureProvider(settings *clientSettings, passwords api.Passwords) (api.AzureProvider, error) {
+func newAzureProvider(settings *clientSettings, passwords api.Passwords) (AzureProvider, error) {
+	// @TODO see if this is still needed
 	// build clients that use the GraphRBAC endpoint
-	userAgent := useragent.PluginString(settings.PluginEnv, userAgentPluginName)
-
-	var appClient api.ApplicationsClient
-	var groupsClient api.GroupsClient
-	var spClient api.ServicePrincipalClient
+	//userAgent := useragent.PluginString(settings.PluginEnv, userAgentPluginName)
+	httpClient := cleanhttp.DefaultClient()
 
-	graphURI, err := api.GetGraphURI(settings.Environment.Name)
+	cred, err := getTokenCredential(settings)
 	if err != nil {
 		return nil, err
 	}
 
-	graphApiAuthorizer, err := getAuthorizer(settings, graphURI)
+	msGraphAppClient, err := api.NewMSGraphApplicationClient(settings.GraphURI, cred)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to create MS graph client: %w", err)
 	}
 
-	msGraphAppClient, err := api.NewMSGraphApplicationClient(settings.SubscriptionID, userAgent, graphURI, graphApiAuthorizer)
+	opts := getClientOptions(settings, httpClient)
+
+	raClient, err := armauthorization.NewRoleAssignmentsClient(settings.SubscriptionID, cred, opts)
 	if err != nil {
 		return nil, err
 	}
 
-	appClient = msGraphAppClient
-	groupsClient = msGraphAppClient
-	spClient = msGraphAppClient
-
-	// build clients that use the Resource Manager endpoint
-	resourceManagerAuthorizer, err := getAuthorizer(settings, settings.Environment.ResourceManagerEndpoint)
+	rdClient, err := armauthorization.NewRoleDefinitionsClient(cred, opts)
 	if err != nil {
 		return nil, err
 	}
 
-	raClient := authorization.NewRoleAssignmentsClientWithBaseURI(settings.Environment.ResourceManagerEndpoint, settings.SubscriptionID)
-	raClient.Authorizer = resourceManagerAuthorizer
-	raClient.AddToUserAgent(userAgent)
+	p := &provider{
+		settings:   settings,
+		httpClient: httpClient,
```

Comment:
I'm wondering if we need to hold onto this here? I see that it is given as an option for the role assignments/definitions clients. But it's unused after that point. Is there any good reason to hold onto it in a field of `provider`?

Reply:
Ah nice, I did not notice that 😅 I don't think there's any reason to tie it to the provider struct for the reasons you mentioned above 👍🏼